File: Programming Regular/Daily_Coding_Problem/645_Problem/problem_645.py
#!/usr/bin/env python3

# --------------------------------------------------

# JLF  -  20201102 - 20201115

#  from Daily Coding Problem: Problem #645 [easy]

#  Good morning! Here's your coding interview problem for today.
#  
#  This problem was asked by Microsoft.
#  
#  Given a 2D matrix of characters and a target word, write a function that returns whether the word can be found in matrix in the matrix by going left-to-right, or up-to-down.
#  
#  For example, given the following matrix:
#  
#  [['F', 'A', 'C', 'I'],
#   ['O', 'B', 'Q', 'P'],
#   ['A', 'N', 'O', 'B'],
#   ['M', 'A', 'S', 'S']]
#  
#  and the target word 'FOAM', you should return true, since it's the leftmost column. Similarly, given the target word 'MASS', you should return true, since it's the last row.
#  
#  

# --------------------------------------------------

def mL():
        print("")
        print("")
        print("------------------------------------------------------")
        print("")
        print("")

# --------------------------------------------------

def two_string_method():
      # First Try:  make a string out of matrix, and an other out of
      # a flipped matrix, then see if the words are in there:

      # make variables to hold the data given in the question:

      print("------------------------------------------------")      
      print("two_string_method")

      matrix = [['F', 'A', 'C', 'I'],['O', 'B', 'Q', 'P'],['A', 'N', 'O', 'B'],['M', 'A', 'S', 'S']]
      target_word_1 = 'FOAM'
      target_word_2 = 'MASS'


      # --------------------------------------------------

      # make one big string from horizontal matrix letters:

      damn_list = []
      for x in range(4):
            for i in range(4):
                  damn_list.extend(matrix[i][x])

      horizontal_letter_string = ''
      horizontal_letter_string = horizontal_letter_string.join(damn_list)


      # --------------------------------------------------

      # make one big string from vertical matrix letters:

      new_list = []
      for i in range(4):
            new_list = new_list + matrix[i]

      vertical_letter_string = ''
      vertical_letter_string = vertical_letter_string.join(new_list)

      # --------------------------------------------------


      # see if the target words are in the matrix, via the 
      # created strings, and if yes show that info on the screen:

      print("")
      print("target word 1 => ", target_word_1)
      if target_word_1 in vertical_letter_string:
            print("found horizontally in matrix")
      if target_word_1 in horizontal_letter_string:
            print("found vertically in matrix")

      print("")
      print("target word 2 => ", target_word_2)
      if target_word_2 in vertical_letter_string:
            print("found horizontally in matrix")
      if target_word_2 in horizontal_letter_string:
            print("found vertically in matrix")

# --------------------------------------------------


def one_big_string_method():

      # make variables to hold the data given in the question:

      matrix = [['F', 'A', 'C', 'I'],['O', 'B', 'Q', 'P'],['A', 'N', 'O', 'B'],['M', 'A', 'S', 'S']]
      target_word_1 = 'FOAM'
      target_word_2 = 'MASS'

      # make one big list from matrix letters:

      print("------------------------------------------------")      
      print("one_big_string_method")

      damn_list = []
      for x in range(4):
            for i in range(4):
                  damn_list.extend(matrix[i][x])
      for i in range(4):
            for x in range(4):
                  damn_list.extend(matrix[i][x])


      # turn damn_list into one big_string

      big_string = ''
      big_string = big_string.join(damn_list)

      print("big_string => ", big_string)

      if target_word_1 in big_string:
            print("found ", target_word_1)
      if target_word_2 in big_string:
            print("found ", target_word_2)
      # --------------------------------------------------


# --------------------------------------------------------

def find_words_as_lists():

      matrix = [['F', 'A', 'C', 'I'],['O', 'B', 'Q', 'P'],['A', 'N', 'O', 'B'],['M', 'A', 'S', 'S']]
      target_word_1 = 'FOAM'
      target_word_2 = 'MASS'

      list1 = [] 
      list2 = [] 
      list1[:0] = (target_word_1)
      list2[:0] = (target_word_2)

      print("------------------------------------------------")      
      print("find_words_as_lists")

      print("We are looking for:")
      print("list1 => ", list1)
      print("list2 => ", list2)
      print("")

      for x in range(4): 
            print("matrix[", x, "] => ", matrix[x])
            if list1 == matrix[x]:
                  print("Found ", list1, "!")
            if list2 == matrix[x]:
                  print("Found ", list2, "!")

      flipped_matrix = []
      temp_list = []

      for y in range(4):
            for x in range(4):
                  temp_list.extend(matrix[x][y])
            flipped_matrix.append(temp_list)
            temp_list = []

      for x in range(4): 
            print("flipped_matrix[", x, "] => ", flipped_matrix[x])
            if list1 == flipped_matrix[x]:
                  print("Found ", list1, "!")
            if list2 == flipped_matrix[x]:
                  print("Found ", list2, "!")

# --------------------------------------------------------

def main():
      mL()
      
      # show the matrix on the screen:

      print("matrix => ")
      print("")
      print(['F', 'A', 'C', 'I'])
      print(['O', 'B', 'Q', 'P'])
      print(['A', 'N', 'O', 'B'])
      print(['M', 'A', 'S', 'S'])

      one_big_string_method()
      two_string_method()
      find_words_as_lists()
      mL()

      
if __name__ == '__main__':
      main()

File: Programming Regular/Daily_Coding_Problem/645_Problem/problem_645.cpp
#include <iostream>
#include <string>
using namespace std;

/*

jlf  -  20201115

#  from Daily Coding Problem: Problem #645 [easy]

Good morning! Here's your coding interview problem for today.
 
This problem was asked by Microsoft.
 
 Given a 2D matrix of characters and a target word, write a function that returns whether the word can be found in     matrix in the matrix by going left-to-right, or up-to-down.
 
 For example, given the following matrix:
 
 [['F', 'A', 'C', 'I'],
  ['O', 'B', 'Q', 'P'],
  ['A', 'N', 'O', 'B'],
  ['M', 'A', 'S', 'S']]
 
 and the target word 'FOAM', you should return true, since it's the leftmost column. Similarly, given the target w    ord 'MASS', you should return true, since it's the last row.
*/ 

void showData() {

      cout << "\nHere is the matrix we are working with:\n";
      cout << "\n['F', 'A', 'C', 'I']";
      cout << "\n['O', 'B', 'Q', 'P']";
      cout << "\n['A', 'N', 'O', 'B']";
      cout << "\n['M', 'A', 'S', 'S']";
      cout << endl; 
      
      cout << "\ntargets words => 'FOAM' and 'MASS' \n";

} // end showData

// ------------------------------------------------------------------------------

void makeLine() {

      cout << "\n\n-----------------------------------------------------\n\n";

} // end of makeLine()

//---------------------------------------------------------------------------------

int main() {

      char matrix[4][4];
      string word3 = "";
      string word1 = "FOAM";
      string word2 = "MASS";
      
      makeLine();

      showData();

      // put the matrix in a data structure:      
      matrix[0][0] = 'F';
      matrix[0][1] = 'A';
      matrix[0][2] = 'C';
      matrix[0][3] = 'I';
      matrix[1][0] = 'O';
      matrix[1][1] = 'B';
      matrix[1][2] = 'Q';
      matrix[1][3] = 'P';
      matrix[2][0] = 'A';
      matrix[2][1] = 'N';
      matrix[2][2] = 'O';
      matrix[2][3] = 'B';
      matrix[3][0] = 'M';
      matrix[3][1] = 'A';
      matrix[3][2] = 'S';
      matrix[3][3] = 'S';

      // make a string from the vertical matrix letters:
      word3 = word3 + matrix[0][0];
      word3 = word3 + matrix[0][1];
      word3 = word3 + matrix[0][2];
      word3 = word3 + matrix[0][3];
      word3 = word3 + matrix[1][0];
      word3 = word3 + matrix[1][1];
      word3 = word3 + matrix[1][2];
      word3 = word3 + matrix[1][3];
      word3 = word3 + matrix[2][0];
      word3 = word3 + matrix[2][1];
      word3 = word3 + matrix[2][2];
      word3 = word3 + matrix[2][3];
      word3 = word3 + matrix[3][0];
      word3 = word3 + matrix[3][1];
      word3 = word3 + matrix[3][2];
      word3 = word3 + matrix[3][3];

      // add to string the horizontal matrix letters:
      word3 = word3 + matrix[0][0];
      word3 = word3 + matrix[1][0];
      word3 = word3 + matrix[2][0];
      word3 = word3 + matrix[3][0];
      word3 = word3 + matrix[0][1];
      word3 = word3 + matrix[1][1];
      word3 = word3 + matrix[2][1];
      word3 = word3 + matrix[3][1];
      word3 = word3 + matrix[0][2];
      word3 = word3 + matrix[1][2];
      word3 = word3 + matrix[2][2];
      word3 = word3 + matrix[3][2];
      word3 = word3 + matrix[0][3];
      word3 = word3 + matrix[1][3];
      word3 = word3 + matrix[2][3];
      word3 = word3 + matrix[3][3];

      cout << endl;
     
      // show string on screen: 
      cout << "Turn matrix into long string made from horizontal and";
      cout << "\nvertical readings of matrix. =>\n";      
      cout << word3;
      cout << "  length of string => " << word3.length() << endl;

      // check to see if word1 is in the string:
      for ( int i=0; i<word3.length()-3; i++ ) 
      {
            if (( word3[i] == word1[0] ) && 
               ( word3[i+1] == word1[1] ) &&
               ( word3[i+2] == word1[2] ) && 
               ( word3[i+3] == word1[3] ))
               {
                  cout << "\nfound a match for " << word1 << " at position [" << i << "]"; 
               }
      }

      // check to see if word2 is in the string:
      for ( int i=0; i<word3.length()-3; i++ ) 
      {
            if (( word3[i] == word2[0] ) && 
               ( word3[i+1] == word2[1] ) &&
               ( word3[i+2] == word2[2] ) && 
               ( word3[i+3] == word2[3] ))
               {
                  cout << "\nfound a match for " << word2 << " at position [" << i << "]"; 
               }
      }

      makeLine();


} // end of main
